addons/l10n_it_edi/models/res_company.py

```diff
@@ -82,20 +82,28 @@ class ResCompany(models.Model):
     @api.constrains('l10n_it_has_eco_index',
                     'l10n_it_eco_index_office',
                     'l10n_it_eco_index_number',
-                    'l10n_it_eco_index_share_capital',
-                    'l10n_it_eco_index_sole_shareholder',
                     'l10n_it_eco_index_liquidation_state')
     def _check_eco_admin_index(self):
         for record in self:
-            if not record.l10n_it_has_eco_index:
-                continue
-            if not record.l10n_it_eco_index_office\
-               or not record.l10n_it_eco_index_number\
-               or not record.l10n_it_eco_index_share_capital\
-               or not record.l10n_it_eco_index_sole_shareholder\
-               or not record.l10n_it_eco_index_liquidation_state:
+            if (record.l10n_it_has_eco_index
+                and (not record.l10n_it_eco_index_office
+                     or not record.l10n_it_eco_index_number
+                     or not record.l10n_it_eco_index_liquidation_state)):
                 raise ValidationError(_("All fields about the Economic and Administrative Index must be completed."))
 
+    @api.constrains('l10n_it_has_eco_index',
+                    'l10n_it_eco_index_share_capital',
+                    'l10n_it_eco_index_sole_shareholder')
+    def _check_eco_incorporated(self):
+        """ If the business is incorporated, both these fields must be present.
+            We don't know whether the business is incorporated, but in any case the fields
+            must be both present or not present. """
+        for record in self:
+            if (record.l10n_it_has_eco_index
+                and bool(record.l10n_it_eco_index_share_capital) ^ bool(record.l10n_it_eco_index_sole_shareholder)):
```

Comment:
not sure about the ^ here (xor?)

Reply:
@lordkrandel Does not it contradict with the docstring?

Reply:
No? 

> in any case the fields must be both present or not present. 

If there's just one, then `ValidationError`

Reply:
Idd, sorry.